### PR TITLE
CUMULUS-1232: Deduplicate duplicate-handling code

### PR DESCRIPTION
**Pull request by mennovandiermen**
**Summary:** Summary of changes

Addresses [CUMULUS-1232: Simplify Sync-Granule](https://bugs.earthdata.nasa.gov/browse/CUMULUS-1232)

## Changes

* Unifies duplicate handling in `ingest/granule.handleDuplicateFile` for maintainability.
* Changed `ingest/granule.ingestFile` and `move-granules/index.moveFileRequest` to use new function.
* Moved file versioning code to `ingest/granule.moveGranuleFileWithVersioning`
* `ingest/granule.verifyFile` now falls back to `fileSize` for verification if checksum is not available in input and throws `UnexpectedFileSize` error for fileSize not matching input.
* `ingest/granule.verifyFile` logs warning if neither checksum nor fileSize are available.
* Add `skip` and `replace` duplicateHandling unit tests.
* Update tests with correct file sizes.

## PR Checklist

- [x] Update CHANGELOG
- [x] Unit tests
- [x] Adhoc testing
- [x] Integration tests

### CODE REVIEW

File: packages/ingest/granule.js [hunk 1]
```diff
@@ -381,7 +363,21 @@ class Granule {
    */
   async verifyFile(file, bucket, key, options = {}) {
     const [type, value] = await this.retrieveSuppliedFileChecksumInformation(file);
-    if (!type || !value) return [null, null];
+    if (!type || !value) {
```
Comment (jennyhliu) [inline packages/ingest/granule.js]: If both checksum and fileSize of the source file are provided, should we check both?

File: packages/ingest/granule.js [hunk 2]
```diff
@@ -690,11 +650,115 @@ function copyGranuleFile(source, target, options) {
...
+* @returns {Promise} returns a promise that is resolved when the file is moved
 **/
 async function moveGranuleFile(source, target, options) {
   await copyGranuleFile(source, target, options);
-  return aws.s3().deleteObject(source).promise();
+  return aws.deleteS3Object(source.Bucket, source.Key);
+}
+
+/**
+* Move granule file from one s3 bucket & keypath to another
```
Comment (jennyhliu) [inline packages/ingest/granule.js]: This function is for moveGranuleFile`WithVersioning`, update function description.

File: packages/ingest/granule.js (hunk 1 above)
Comment (mennovandiermen) [reply packages/ingest/granule.js]: I used `fileSize` as a fallback to have at least a minimal check for a file without a provided checksum, as opposed to silently returning `null, null`. I'm not sure that a file with a matching checksum but a non-matching fileSize is a major enough edge case to check both. `fileSize` is generally more unreliable as a verifying factor so I'm not eager to rely on it, but I'm not really convinced which is the better move either way. Any specific thoughts on why or why not?

File: packages/ingest/test/test-granule.js [hunk 3]
```diff
@@ -666,6 +666,158 @@ test('ingestFile throws error when configured to handle duplicates with error',
...
+  params.Body = randomString(100);
+  await s3PutObject(params);
+  const newfiles = await testGranule.ingestFile(file, destBucket, duplicateHandling);
+  t.is(newfiles.length, 1);
+  t.true(newfiles[0].duplicate_found);
+  t.is(newfiles[0].fileSize, oldfiles[0].fileSize);
+  t.not(newfiles[0].fileSize, params.Body.length);
+});
+
+test('ingestFile replaces file when duplicateHandling is replace', async (t) => {
```
Comment (ifestus) [inline packages/ingest/test/test-granule.js]: Correct me if I'm wrong, but this is testing the case where no `cksum` or `filesize` are provided. Should there be another test that validates the functionality of a file with the same `cksum` with `replace` duplicate handling being left alone?

File: packages/ingest/granule.js [hunk 4]
```diff
@@ -690,11 +650,115 @@ function copyGranuleFile(source, target, options) {
...
+  checksumFunction,
+  syncFileFunction
+}) {
+  if (duplicateHandling === 'error') {
+    // Have to throw DuplicateFile and not WorkflowError, because the latter
+    // is not treated as a failure by the message adapter.
+    throw new errors.DuplicateFile(`${target.Key} already exists in ${target.Bucket} bucket`);
+  } else if (duplicateHandling === 'version') {
+    // sync to staging location if required
+    if (syncFileFunction) await syncFileFunction(source.Bucket, source.Key);
```
[the remark this answers is not in the document]
Comment (mennovandiermen) [reply packages/ingest/granule.js]: Correct.

File: packages/ingest/granule.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mennovandiermen) [reply packages/ingest/granule.js]: Correct, because we need both a `type` and `value` to verify a checksum, unless we want to assume that a checksum without a type is a `cksum` checksum. Our checksum functions support algorithms other than `cksum`, and some providers use `md5`. CNM for example assumes that the checksumType is `md5`. I will improve the logged statement.

File: packages/ingest/granule.js [hunk 5]
```diff
@@ -472,99 +468,63 @@ class Granule {
...
-    // from input file
-    return renamedFiles.concat({ Bucket: bucket, Key: destinationKey }).map((f) => {
-      if (f.Key === destinationKey) return stagedFile;
-      return {
+    // Set final filesize
+    stagedFile.fileSize = await aws.getObjectSize(destinationBucket, destinationKey);
+    // return all files, the renamed files don't have the same properties
+    // (name, fileSize, checksum) as input file
+    log.debug(`returning ${JSON.stringify(stagedFile)}`);
+    return [stagedFile].concat(versionedFiles.map((f) => (
```
[the remark this answers is not in the document]
Comment (mennovandiermen) [reply packages/ingest/granule.js]: It's important to note the return value from handleDuplicateFiles, which for all cases other than `version` is `[]`. This means that `[stagedFile].concat()` will concatenate an empty array, so it will return an array with only the staged file. Only for the `version` case does it return all versioned filenames.

File: packages/ingest/test/test-granule.js (hunk 3 above)
Comment (mennovandiermen) [reply packages/ingest/test/test-granule.js]: Checksumming only occurs in the `version` case, not the `replace` case, as there is no 'duplicate checksum' AC on the `replace` ticket: https://bugs.earthdata.nasa.gov/browse/CUMULUS-781

File: packages/ingest/granule.js [hunk 6]
```diff
@@ -472,99 +468,63 @@ class Granule {
...
     const destinationKey = path.join(stagingPath, file.name);
 
-    const s3ObjAlreadyExists = await aws.s3ObjectExists({
-      Bucket: bucket,
-      Key: destinationKey
-    });
-
     // the staged file expected
-    const stagedFile = Object.assign(file,
+    const stagedFile = Object.assign(cloneDeep(file),
```
[the remark this answers is not in the document]
Comment (mennovandiermen) [reply packages/ingest/granule.js]: New unit tests caught it for me, actually.

File: packages/ingest/test/test-granule.js (hunk 3 above)
Comment (ifestus) [reply packages/ingest/test/test-granule.js]: You're right - I transposed `replace` and `version` in my head.

File: packages/ingest/granule.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (mennovandiermen) [reply packages/ingest/granule.js]: There is probably an underlying assumption that the duplicateHandling will not change once a granule has been ingested here, and that may be a wrong assumption, but in the interest of only refactoring to simplify, I have maintained current inputs & outputs. If you think we should address that by, say, returning any versioned filenames for a `replace` granule, I'm not inclined to disagree with you, but I do think that would be out of scope for this ticket/PR.

File: packages/ingest/test/test-granule.js (hunk 3 above)
Comment (mennovandiermen) [reply packages/ingest/test/test-granule.js]: For the version case, you're right that there is no checksumming performed in the unit test. It is something we cover extensively in integration tests. I'd like to have a discussion about whether to move some of that to unit tests (out of scope for this ticket imo), but I don't know that we need to test it again in these unit tests.

Comment (ifestus) [reply packages/ingest/test/test-granule.js]: Speaking of which, I don't see any test of that sort for `version`. Do you think that would be something worthwhile to add as a unit test?

File: packages/ingest/granule.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (ifestus) [reply packages/ingest/granule.js]: I think this is good. I bring this up because I want to make sure that changing this value doesn't break the assumption that CNM messages have that [cksum type is optional and defaults to cksum](https://wiki.earthdata.nasa.gov/display/CUMULUS/Cloud+Notification+Mechanism#CloudNotificationMechanism-ResponseMessageFields#granule-fields).

File: packages/ingest/granule.js (hunk 2 above)
Comment (mennovandiermen) [reply packages/ingest/granule.js]: Good catch!

File: packages/ingest/granule.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mennovandiermen) [reply packages/ingest/granule.js]: The CNM default is actually `it is assumed to be md5.`, but their CNM translation lambda should hand populating the `checksumType` field with that default value if it does not exist in the original CNM message.

[the remark this answers is not in the document]
Comment (mennovandiermen) [reply packages/ingest/granule.js]: (At least it will if they accept my PR to that function. Yay insider knowledge!)

File: packages/ingest/granule.js (hunk 2 above)
Comment (mennovandiermen) [reply packages/ingest/granule.js]: c411cad 👍

File: packages/ingest/granule.js (hunk 1 above)
Comment (jennyhliu) [reply packages/ingest/granule.js]: I don't have strong opinion on this.  But since fileSize mismatch is an error when checksum is not present, it should be an error as well if checksum is present.

Comment (mennovandiermen) [reply packages/ingest/granule.js]: I've made the change to always check fileSize as I could not think of a good reason not to.
